### PR TITLE
Refactor CSRMatrix

### PR DESCRIPTION
**Pull request by angriman**
This adds a template parameter to the `CSRMatrix` class allowing to store generic data types. The class is renamed to `CSRGeneralMatrix`. Additional minor changes:
- Remove ad-hoc implementation of binary search in favor of `std::lower_bound`
- Add `noexcept` operator where applicable
- Replace `push_back` with `emplace_back` where applicable (and addition of two new constructors for `Triplet`)

### CODE REVIEW

File: include/networkit/algebraic/CSRGeneralMatrix.hpp
Comment (avdgrinten) [inline include/networkit/algebraic/CSRGeneralMatrix.hpp]: Add yourself as an author (in addition to Michael Wegner).

File: include/networkit/algebraic/CSRMatrix.hpp
Comment (avdgrinten) [inline include/networkit/algebraic/CSRMatrix.hpp]: We should not delete the entire header file. This requires quite a bit of refactoring in NetworKit and it might also require users to adopt. Instead, keep `CSRMatrix.hpp` header, let it include the `CSRGeneralMatrix.hpp` header and move the typedef to the `CSRMatrix.hpp` header.

File: include/networkit/algebraic/CSRGeneralMatrix.hpp
Comment (avdgrinten) [inline include/networkit/algebraic/CSRGeneralMatrix.hpp]: Indentation is broken here (and in large parts of the remainder of the file). Please fix it and make it consistent.

File: include/networkit/algebraic/CSRGeneralMatrix.hpp
Comment (avdgrinten) [inline include/networkit/algebraic/CSRGeneralMatrix.hpp]: Move the the initializer list to its own line:
```cpp
CSRGeneralMatrix()
: rowIdx(0), columnIdx(0), nonZeros(0), nRows(0), nCols(0), isSorted(true), zero(0) {}
```

File: include/networkit/algebraic/CSRGeneralMatrix.hpp
Comment (avdgrinten) [inline include/networkit/algebraic/CSRGeneralMatrix.hpp]: Actually, looking at the code, there are still a lot of places, where `double` is used. This does not look intentional?

Comment (angriman) [reply include/networkit/algebraic/CSRGeneralMatrix.hpp]: My bad, it is not intentional, I'll change them to `ValueType`.

> LGTM now. I did not review the algorithm; I assume that they did not change except for changing `double` -> `ValueType` (?).

Exactly, apart from that there are no further changes.

**Comment by avdgrinten** [comment]
Why does the code break with an earlier version of Clang? We should not increase the version of Clang without a good reason. People who want run NetworKit on clusters that they do not control might not have the option to upgrade Clang.

**Comment by angriman** [comment]
The build fails with clang 3.8 because of this error:
```
/home/travis/build/networkit/networkit/include/networkit/algebraic/CSRGeneralMatrix.hpp:1231:10: error: 
      definition with same mangled name as another definition
    void forNonZeroElementsInRow(index row, L handle) const {
         ^
/home/travis/build/networkit/networkit/include/networkit/algebraic/CSRGeneralMatrix.hpp:1231:10: note: 
      previous definition is here
```

I searched for this error and did not find much apart [this](https://bugs.llvm.org/show_bug.cgi?id=33924) page. It could be a bug in clang 3.8 but I am not sure about that. Do you have some other solutions in mind?

**Comment by angriman** [comment]
Moving `forNonZeroElementsInRow` somewhere else in the class did not solve the issue, I still get the same error. Shall we switch to clang 3.9?

**Comment by angriman** [comment]
Now that we raised the minimum supported clang version to 3.9 this PR can be reviewed.

**Review by fabratu — DISMISSED** [review]
LGTM.

**Comment by angriman** [comment]
We also need approval from @avdgrinten.

File: include/networkit/algebraic/CSRGeneralMatrix.hpp
Comment (avdgrinten) [inline include/networkit/algebraic/CSRGeneralMatrix.hpp:58]: We do we use an ad-hoc quicksort instead of using the standard library?

**Review by avdgrinten — COMMENTED** [review]
I have one comment below.

As a side note: a few algorithms are *really* inefficient here, for example setValue(). There is no way to implement setValue() well for CSR, so why does the class provide it at all?

**Comment by angriman** [comment]
I agree that there are several inefficiencies in this class that should be fixed, I already fixed a trivial one in the last commit. However, I think that fixing those inefficiencies is out of the scope of this PR and it should be done on a separate PR.

**Comment by hmeyerhenke** [comment]
Not sure that setValue() is really obsolete. But the docs should be clear about the inherent inefficiency. Other -- unnecessary -- inefficiencies should be addressed. This is rather old code that has not been touched for quite some time, I believe.

**Comment by hmeyerhenke** [comment]
But I leave it up to you guys if they are addressed in this PR or separately.

**Comment by avdgrinten** [comment]
It's not necessarily the case that the code is inefficient due to the implementation but it just provides tons of operations that you really would not use if you want your code to be fast (setting values, getting individual values, extracting columns, etc.). These operations are simply not fast because CSR cannot implement them in a fast way.

**Comment by hmeyerhenke** [comment]
I know @avdgrinten. Such operations should not be the preferred way to do things, I completely agree. But there may be cases where they are simply necessary. We can make a decision, though, that we do not want to support such cases. But that's a decision I'd like to be involved in and should follow from a comprehensive discussion of use cases.

**Comment by angriman** [comment]
It might be a good idea to open a new issue/discussion about that.

**Comment by angriman** [comment]
Are there any other issues that needs to be addressed in this PR?

File: include/networkit/algebraic/CSRGeneralMatrix.hpp
Comment (angriman) [reply include/networkit/algebraic/CSRGeneralMatrix.hpp:58]: A possible reason for this is because we need to sort two vectors at once (columnIdx and nonZero). We could replace this ad-hoc quicksort by computing the permutation of indices using `std::sort` and then apply the permutation to the two vectors as done in #734 (but to do so we might want to merge the `applyPermutation` function in #734 first).
Do you know if there is a better way to do this?

File: include/networkit/algebraic/CSRGeneralMatrix.hpp
Comment (avdgrinten) [inline include/networkit/algebraic/CSRGeneralMatrix.hpp:282]: `inline` is redundant if a method is defined within `class { ... }` or `struct { ... }`.

File: include/networkit/algebraic/Vector.hpp
Comment (avdgrinten) [inline include/networkit/algebraic/Vector.hpp:312]: Unrelated change?

File: include/networkit/numerics/LAMG/Lamg.hpp
Comment (avdgrinten) [inline include/networkit/numerics/LAMG/Lamg.hpp:151]: Unrelated change?

File: include/networkit/numerics/LAMG/MultiLevelSetup.hpp
Comment (avdgrinten) [inline include/networkit/numerics/LAMG/MultiLevelSetup.hpp:347]: Unrelated change?

File: networkit/cpp/algebraic/DynamicMatrix.cpp
Comment (avdgrinten) [inline networkit/cpp/algebraic/DynamicMatrix.cpp:249]: Unrelated change?

File: networkit/cpp/io/MatrixMarketReader.cpp
Comment (avdgrinten) [inline networkit/cpp/io/MatrixMarketReader.cpp:108]: Unrelated change?

File: networkit/cpp/numerics/LAMG/MultiLevelSetup.cpp
Comment (avdgrinten) [inline networkit/cpp/numerics/LAMG/MultiLevelSetup.cpp:120]: Unrelated change?

File: networkit/cpp/viz/PivotMDS.cpp
Comment (avdgrinten) [inline networkit/cpp/viz/PivotMDS.cpp:38]: Unrelated change?

File: include/networkit/algebraic/Vector.hpp
Comment (angriman) [reply include/networkit/algebraic/Vector.hpp:312]: If I do not change this then the clang-tidy build fails due to `modernize-use-emplace`.

Comment (avdgrinten) [reply include/networkit/algebraic/Vector.hpp:312]: Why doesn't clang-tidy fail right now?

**Comment by angriman** [comment]
I addressed the comments, the `emplace_back` were not really necessary, so I removed them.

**Review by avdgrinten — APPROVED** [review]
LGTM now.

**Pull request merged**
